idx/metric_definition.go

```diff
@@ -0,0 +1,546 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+	"github.com/tinylib/msgp/msgp"
+)
+
+//go:generate msgp
+
+func init() {
+	msgp.RegisterExtension(95, func() msgp.Extension { return &MetricName{} })
+	msgp.RegisterExtension(90, func() msgp.Extension { return &TagKeyValues{} })
+}
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
```

Comment:
functions that aren't called from other packages, or shouldn't be called from other packages, should be unexported

Reply:
It already has a field named `nodes`. I can't rename it to `nodes()` but we could use something like `getNodes()`. I think `getNodes()` doesn't follow Go best practices though. The other option would be to rename the `nodes` field to something else.

Reply:
oh then we can just remove the method and access the property directly

Reply:
We can't because it is accessed [in idx to release interned items](https://github.com/grafana/metrictank/blob/add-tag-interning/idx/idx.go#L150). I don't remember why I needed to make it unexported in the first place though. I will explore that again instead. I think it had something to do with msgp marshaling.

Reply:
you can make msgp ignore an exported property by appending `msg:"-"` to it's definition in the type definition